### PR TITLE
[OCPCLOUD-1107] External cloud-provider support via FeatureGate in post-install

### PR DESCRIPTION
**Pull request by Danil-Grigorev**
**- What I did**

Implementing external cloud provider selection forced by feature gate. Described in the proposal: https://github.com/openshift/enhancements/pull/463/files#diff-3e0e2c48e70215076dfe36c13768a823ab7080d929d80292f37db2ef5a2121e8R201

This PR is main blocker on other work with integrating CCM in Openshift: https://issues.redhat.com/browse/OCPCLOUD-1107
 
**- How to verify it**

**- Description for the changelog**

### CODE REVIEW

File: pkg/controller/kubelet-config/kubelet_config_controller.go
Comment (kikisdeliveryservice) [inline pkg/controller/kubelet-config/kubelet_config_controller.go:325]: macherrors-> matcherrors

**Review by kikisdeliveryservice — COMMENTED** [review]
feel like the team/arch also needs to review the enhancement 

/hold

**Comment by kikisdeliveryservice** [comment]
Feels like this should leverage: https://github.com/openshift/machine-config-operator/pull/2352  via: https://github.com/openshift/api/blob/master/config/v1/types_feature.go#L121

File: pkg/controller/template/render.go
Comment (Danil-Grigorev) [reply pkg/controller/template/render.go:342]: Eventually, yes. In TechPreviewNoUpgrade it is on feature gate to set invariant configuration when found. I don't think anyone will risk setting such gate on production cluster for platform we don't have configs for. This just means less changes in development cycle, if we don't make it a switch.

Comment (JoelSpeed) [reply pkg/controller/template/render.go:342]: I had thought we would want this to only affect things on platforms we have allowed it to, hence being provider specific, let's discuss on the call next week

**Review by kikisdeliveryservice — COMMENTED** [review]
@Danil-Grigorev I've gone through the enhancements, but am having problems connecting the dots here.

Could you add some detail/specific description as to what this PR is trying to accomplish, whether this PR is like step 1 of X etc... if this is day2 post-install is this user initiated? etc? Will there be any followup prs or are there prereq prs for this? etc

Thanks! :)

**Comment by Danil-Grigorev** [comment]
@kikisdeliveryservice Sure. The prerequisite PR will be an API change for FeatureGate in https://github.com/openshift/api/pull/738 The name of the featureGate, it's location and related changes in functionality may be corrected based on new data from https://github.com/kubernetes/enhancements/pull/2443. What it will result into - we will have a way to deploy a techPreview clusters with out-of-tree providers code, and attempt migration from in-tree in post install time, while identifying issues (CI use-case). It is user initiated.

Ideal scenario - follow up PRs will only remove selections from platform switch case, say: AWS is now by default `external` in 1.24. Realistically, we will still have to force some `kubelet` flags even after the upstream feature gate is implemented, as the goal is to identify issues in migration, which may break some storage, networking and cloud-credentials functionality in the cluster if done blindly.

Actual change, which we will probably need - implement our selection (or more realistically wait for some helper upstream implementation) in MCO install phase, where the bootstrap methods are called. This way a fresh cluster will get out-of-tree providers by default with related `DisableCloudProviders` feature gate . This is out-of-scope for this PR.

**Comment by kikisdeliveryservice** [comment]
Thanks for the summary @Danil-Grigorev !

File: pkg/controller/kubelet-config/kubelet_config_features.go
Comment (JoelSpeed) [inline pkg/controller/kubelet-config/kubelet_config_features.go:205]:
```suggestion
func generateFeatureMap(features *osev1.FeatureGate, exclusions ...string) (*map[string]bool, error) {
```

File: pkg/controller/kubelet-config/kubelet_config_features.go
Comment (JoelSpeed) [inline pkg/controller/kubelet-config/kubelet_config_features.go:229]:
```suggestion
	for _, excluded := range exclusions {
```

File: pkg/controller/template/template_controller.go
Comment (JoelSpeed) [inline pkg/controller/template/template_controller.go:191]: This log line is no longer correct if this is being called from `updateFeature` for example, does this need to be more generic?
```suggestion
	glog.V(4).Infof("Re-syncing ControllerConfig %s due to dependency change", cfg.Name)
```

File: pkg/controller/template/template_controller.go
Comment (JoelSpeed) [inline pkg/controller/template/template_controller.go:225]: Do we not need to enqueue a reconcile if the feature is deleted?

File: pkg/controller/template/template_controller.go
Comment (JoelSpeed) [inline pkg/controller/template/template_controller.go:478]: Why not wrap the error before logging it? Gives more context right?

File: pkg/controller/template/render.go
Comment (Danil-Grigorev) [reply pkg/controller/template/render.go:337]: Because of the deprecation warning. library-go still uses infra.status.platform, while is recommended to use infra.status.platformStatus.type now. This causes `verify` check to fail.

File: pkg/controller/template/render.go
Comment (Danil-Grigorev) [reply pkg/controller/template/render.go:342]: I'm only replicating usual error check with skipping the logic if is found. Will rather leave it as is.

File: pkg/controller/template/template_controller.go
Comment (Danil-Grigorev) [reply pkg/controller/template/template_controller.go:225]: No, it is usual for such static resources to be re-queued only on add/update. Removing it and reconciling is as if you rip off a part of cluster, and therefor destroying it's health.

Comment (JoelSpeed) [reply pkg/controller/template/template_controller.go:225]: This affects the state of the output of MCO though, so deleting it does infer a different desired state, is something going to automatically put it back if you delete it?  I'll leave this decision to the MCO owners but to me it seems natural to reconcile this on delete as well

File: pkg/controller/template/template_controller.go
Comment (JoelSpeed) [inline pkg/controller/template/template_controller.go:188]: Missed this one, also needs an update to be non-secret specific here too
```suggestion
		utilruntime.HandleError(fmt.Errorf("couldn't get ControllerConfig on dependency callback %#v", err))
```

File: pkg/controller/common/constants.go
Comment (kikisdeliveryservice) [inline pkg/controller/common/constants.go:22]: singleton?

File: pkg/controller/template/render.go
Comment (kikisdeliveryservice) [inline pkg/controller/template/render.go:339]: is there a reason you're not returning the error here?

File: pkg/controller/template/render.go
Comment (kikisdeliveryservice) [inline pkg/controller/template/render.go:384]: same as above - does only logging this error make sense?

File: pkg/controller/template/template_controller.go
Comment (kikisdeliveryservice) [inline pkg/controller/template/template_controller.go:188]: the only thing is that changing this string makes the error much more generic

**Review by kikisdeliveryservice — COMMENTED** [review]
made a pass through with some questions/comments :+1:

File: pkg/controller/template/render.go
Comment (Danil-Grigorev) [reply pkg/controller/template/render.go:339]: Yes, the error check is made solely for featureGate validation, and is backed by the filtering behavior, which will return `false`. We don't want to break existing `cloudProvider` workflow, if the featureGate configuration is invalid. This is temporary solution: https://github.com/openshift/library-go/pull/993/files#diff-478af36e9fb994fc80d37b7d2f6ae207c67d8c43b94f98f6ae3e420808958ba9R41 Here is the place where the error comes from, it could be safely ignored and only printed.

File: pkg/controller/template/render.go
Comment (Danil-Grigorev) [reply pkg/controller/template/render.go:384]: https://github.com/openshift/machine-config-operator/pull/2386#discussion_r587561388

File: pkg/controller/template/render.go
Comment (Danil-Grigorev) [reply pkg/controller/template/render.go:339]: It will result in `external` false, which will not break and reconfigure cluster.

Comment (kikisdeliveryservice) [reply pkg/controller/template/render.go:339]: thanks!

File: pkg/controller/template/template_controller.go
Comment (JoelSpeed) [reply pkg/controller/template/template_controller.go:188]: Could pass the type into this function as a string and format string it into the error? Bit of an odd pattern though maybe?

**Comment by Danil-Grigorev** [comment]
@kikisdeliveryservice Latest enhancement update contains detailed information about flags which this PR is setting. Could you give it another review? https://github.com/openshift/enhancements/pull/463/commits/af7b324b8303bb43d80fc6296ca5f4bdeaecb9ef We can't test how CCM initialize node for provider without this patch. Everything is described in the proposal

**Comment by Danil-Grigorev** [comment]
@rphillips @kikisdeliveryservice @JoelSpeed Could you guys give it a review and lablels please? We need it to introduce operator in 4.9, so once the branch opens this should be merged first thing.

File: templates/common/aws/units/afterburn-hostname.service.yaml
Comment (JoelSpeed) [inline templates/common/aws/units/afterburn-hostname.service.yaml:1]: I thought this was being covered in a separate PR? We should keep the scope of this PR just to the feature gate changes and have a separate PR for the AWS hostname issue

**Review by JoelSpeed — COMMENTED** [review]
LGTM, though I think we will want to address the afterburn issue in a separate PR

File: templates/common/aws/units/afterburn-hostname.service.yaml
Comment (Danil-Grigorev) [reply templates/common/aws/units/afterburn-hostname.service.yaml]: Yes it was, although this implementation is the only one which worked. An issue is still there in the other PR and there is no way to fix it: https://github.com/openshift/machine-config-operator/pull/2401#issuecomment-789739611

The implementation does not belong here, was a temporary fix for testing. Removed the last commit.

**Review by JoelSpeed — COMMENTED** [review]
There's a couple of files for group generation which I think are part of a code generation tool, that seem to have changed permissions, this seems unrelated to this PR, have you run `go mod tidy` and `go mod vendor`? I'm surprised the permissions would be changing here

File: pkg/controller/template/template_controller.go
Comment (kikisdeliveryservice) [inline pkg/controller/template/template_controller.go:124]: I keep coming back to removing this secret change logging... Can you leave the log line and maybe with a reword for it to make sense since there's another log line in the func called ?

Comment (kikisdeliveryservice) [reply pkg/controller/template/template_controller.go:124]: As an aside: @sinnykumari should this logging really be v4? on secret change maybe we want it to be more visible? is there an event for this somewhere?

File: pkg/controller/template/template_controller.go
Comment (kikisdeliveryservice) [inline pkg/controller/template/template_controller.go:199]:
```suggestion
		glog.V(4).Infof("Updating Feature: %s", newFeature.Name)
```

File: pkg/controller/template/template_controller.go
Comment (kikisdeliveryservice) [inline pkg/controller/template/template_controller.go:206]:
```suggestion
	glog.V(4).Infof("Adding Feature: %s", features.Name)
```

File: pkg/controller/template/template_controller.go
Comment (kikisdeliveryservice) [inline pkg/controller/template/template_controller.go:224]:
```suggestion
	glog.V(4).Infof("Deleting Feature %s", features.Name)
```

It hasn't happened yet right?

**Review by kikisdeliveryservice — COMMENTED** [review]
A few smaller comments but same q as joel on the permission change

File: pkg/controller/template/render.go
Comment (kikisdeliveryservice) [reply pkg/controller/template/render.go:337]: double checking: is this still true?

**Comment by kikisdeliveryservice** [comment]
`level=error msg=Cluster operator ingress Degraded is True with IngressDegraded: The "default" ingress controller reports Degraded=True: DegradedConditions: One or more other status conditions indicate a degraded state: PodsScheduled=False (PodsNotScheduled: Some pods are not scheduled: Pod "router-default-794c6d77d-msh47" cannot be scheduled: 0/4 nodes are available: 1 node(s) didn't match pod affinity/anti-affinity rules, 1 node(s) didn't match pod anti-affinity rules, 3 node(s) had taint {node-role.kubernetes.io/master: }, that the pod didn't tolerate. Make sure you have sufficient worker nodes.)`

/test e2e-agnostic-upgrade

**Comment by kikisdeliveryservice** [comment]
Would really like to see a successful upgrade run on this but there seem to be a bunch of apimachinery failures.

File: pkg/controller/template/template_controller.go
Comment (sinnykumari) [reply pkg/controller/template/template_controller.go:124]: yes, there is no confidential information we are logging here. So, it makes sense to log it without verbosity.

File: pkg/controller/template/render.go
Comment (Danil-Grigorev) [reply pkg/controller/template/render.go:337]: I guess not, we could use platformStatus.type, let me update this quickly. MCO does bootstrap part after cluster-config-operator, so it should work.

Comment (JoelSpeed) [reply pkg/controller/template/render.go:337]: PlatformStatus.Type is default from OCP 4.2 so I'd expect it to be the one we use here anyway

File: pkg/controller/template/template_controller.go
Comment (Danil-Grigorev) [reply pkg/controller/template/template_controller.go:124]: Added the log line without verbosity

**Comment by Danil-Grigorev** [comment]
@JoelSpeed @kikisdeliveryservice All comments addressed. I checked, permission change happens on revendor, and it seems expected. Please review.

**Review by JoelSpeed — COMMENTED** [review]
/lgtm

Thanks @Danil-Grigorev

**Comment by elmiko** [comment]
just wanted to note that i have been test driving this patch for the last few days and it seems to be working for me. i have not given a thorough testing of the negative case (eg when the gate is not applied), but i have done a basic smoke test on that.

**Comment by elmiko** [comment]
@Danil-Grigorev , i talked with @kikisdeliveryservice in slack and it sounds like we could help here by adding a must-gather for a cluster that has had the feature gate enabled as a day 2 operation. so, basically:
1. create normal cluster with in-tree providers
2. enable feature gate and switch cluster to out-of-tree
3. once cluster is healthy, collect must-gather

i think we are already capturing this behavior in the [openshift-e2e-aws-ccm](https://github.com/openshift/release/blob/master/ci-operator/step-registry/openshift/e2e/aws/ccm/openshift-e2e-aws-ccm-workflow.yaml) step. but since we need this PR to make it work, [none of those jobs have been passing](https://prow.ci.openshift.org/job-history/gs/origin-ci-test/pr-logs/directory/pull-ci-openshift-cluster-cloud-controller-manager-operator-master-e2e-aws-ccm). @Danil-Grigorev does that make sense?

**Comment by kikisdeliveryservice** [comment]
just to follow-on the above, would this ever be disabled day2 @Danil-Grigorev ?

**Comment by Danil-Grigorev** [comment]
@kikisdeliveryservice No, applying the feature gate would move cluster to TechPreview state, where we officially do not support rollback. Yet the code does not prevent it from happening and works both ways.

**Comment by elmiko** [comment]
quick update here, we have shared a must-gather with @kikisdeliveryservice

**Review by kikisdeliveryservice — APPROVED** [review]
Reviewed must gather and verified:

we started off with --cloud-provider=aws
and
successfully switched to --cloud-provider=external

Based on my understanding from: https://github.com/openshift/enhancements/pull/463/files#diff-3e0e2c48e70215076dfe36c13768a823ab7080d929d80292f37db2ef5a2121e8R270 also saw: ExternalCloudProvider: true alongside --cloud-provider=aws which helps force upgrade from intree to out of tree.

Since this has also been extensively tested by that team, let's merge this to unblock work and move forward.

**Pull request merged**
